src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs

```diff
@@ -0,0 +1,211 @@
+
+namespace Microsoft.CodeAnalysis.Remote.Telemetry
+{
+    /// <summary>
+    /// Creates an <see cref="IIncrementalAnalyzer"/> that collects Api usage information from metadata references
+    /// in current solution.
+    /// </summary>
+    [ExportIncrementalAnalyzerProvider(nameof(ApiUsageIncrementalAnalyzerProvider), new[] { WorkspaceKind.RemoteWorkspace }), Shared]
+    internal sealed class ApiUsageIncrementalAnalyzerProvider : IIncrementalAnalyzerProvider
+    {
+        [ImportingConstructor]
+        public ApiUsageIncrementalAnalyzerProvider()
+        {
+        }
+
+        public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
+        {
+            return new Analyzer();
+        }
+
+        private sealed class Analyzer : IIncrementalAnalyzer
+        {
+            // maximum number of symbols to report per project.
+            private const int Max = 2000;
+
+            private const string EventName = "vs/compilers/api";
+            private const string PropertyName = "vs.compilers.api.pii";
+
+            private readonly HashSet<ProjectId> _reported = new HashSet<ProjectId>();
+
+            public void RemoveProject(ProjectId projectId)
+            {
+                lock (_reported)
+                {
+                    _reported.Remove(projectId);
+                }
+            }
+
+            public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
+            {
+                lock (_reported)
+                {
+                    // to make sure that we don't report while solution load, we do this heuristic.
+                    // if the reason we are called is due to "document being added" to project, we wait for next analyze call.
+                    // also, we only report usage information per project once.
+                    // this telemetry will only let us know which API ever used, this doesn't care how often/many times an API
+                    // used. and this data is approximation not precise information. and we don't care much on how many times
+                    // APIs used in the same solution. we are rather more interested in number of solutions or users APIs are used.
+                    if (reasons.Contains(PredefinedInvocationReasons.DocumentAdded) ||
+                        _reported.Contains(project.Id))
+                    {
+                        return;
+                    }
+                }
+
+                var metadataSymbolUsed = new HashSet<ISymbol>();
+                foreach (var document in project.Documents)
+                {
+                    var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                    foreach (var operation in GetOperations(model, cancellationToken))
```

Comment:
so, uh, this feels somewhat concerning from a GC perspective.   this is a walk of all docs, and it produces binding/semantics for everything.  in that sense it's like FSA.  Is this ok because we're stopping once we report for a project, and we're tryin to cap the total number of metadata symbols?

Reply:
this runs on OOP so we don't concern on allocation much.